File: kotlin/scrabble-score/src/main/kotlin/ScrabbleScore.kt
object ScrabbleScore {
    private fun scoreLetter(letter: Char) = when (letter.uppercaseChar()) {
        'Q', 'Z' -> 10
        'J', 'X' -> 8
        'K' -> 5
        'F', 'H', 'V', 'W', 'Y' -> 4
        'B', 'C', 'M', 'P' -> 3
        'D', 'G' -> 2
        else -> 1
    }

    fun scoreWord(word: String) = word.sumOf(::scoreLetter)
}

File: fsharp/scrabble-score/ScrabbleScore.fs
module ScrabbleScore

let private lettersToScore = 
    [("AEIOULNRST", 1); ("DG", 2); ("BCMP", 3); ("FHVWY", 4); ("K", 5); ("JX", 8); ("QZ", 10)] 
    |> Seq.collect (fun (letters, score) -> Seq.map (fun letter -> letter, score) letters)
    |> Map.ofSeq

let scoreLetter letter = Map.tryFind letter lettersToScore |> Option.defaultValue 0

let score (word:string) = Seq.sumBy scoreLetter (word.ToUpperInvariant()) 
